flink-connector-aws/flink-connector-dynamodb/src/test/java/org/apache/flink/connector/dynamodb/source/util/DynamoDbStreamsClientProvider.java

```diff
@@ -0,0 +1,193 @@
+
+import software.amazon.awssdk.awscore.exception.AwsServiceException;
+import software.amazon.awssdk.core.exception.SdkClientException;
+import software.amazon.awssdk.services.dynamodb.model.DescribeStreamRequest;
+import software.amazon.awssdk.services.dynamodb.model.DescribeStreamResponse;
+import software.amazon.awssdk.services.dynamodb.model.GetRecordsRequest;
+import software.amazon.awssdk.services.dynamodb.model.GetRecordsResponse;
+import software.amazon.awssdk.services.dynamodb.model.GetShardIteratorRequest;
+import software.amazon.awssdk.services.dynamodb.model.GetShardIteratorResponse;
+import software.amazon.awssdk.services.dynamodb.model.Shard;
+import software.amazon.awssdk.services.dynamodb.model.StreamDescription;
+import software.amazon.awssdk.services.dynamodb.model.StreamStatus;
+import software.amazon.awssdk.services.dynamodb.streams.DynamoDbStreamsClient;
+import software.amazon.awssdk.services.dynamodb.streams.DynamoDbStreamsServiceClientConfiguration;
+
+import java.util.ArrayDeque;
+import java.util.Deque;
+import java.util.List;
+import java.util.function.Consumer;
+
+/** Provides {@link DynamoDbStreamsClient} with mocked DynamoDbStreams behavior. */
+public class DynamoDbStreamsClientProvider {
+
+    /**
+     * An implementation of the {@link DynamoDbStreamsClient} that allows control over DynamoDb
+     * Service responses.
+     */
+    public static class TestingDynamoDbStreamsClient implements DynamoDbStreamsClient {
+
+        private Deque<DescribeStreamItem> describeStreamItems = new ArrayDeque<>();
+        private Deque<String> shardIterators = new ArrayDeque<>();
+        private Consumer<GetShardIteratorRequest> getShardIteratorValidation;
+        private GetRecordsResponse getRecordsResponse;
+        private Consumer<GetRecordsRequest> getRecordsValidation;
+        private boolean closed = false;
+
+        @Override
+        public String serviceName() {
+            return "dynamodb";
+        }
+
+        @Override
+        public void close() {
+            closed = true;
+        }
+
+        public boolean isClosed() {
+            return closed;
+        }
+
+        public void setNextShardIterator(String shardIterator) {
+            this.shardIterators.add(shardIterator);
+        }
+
+        public void setShardIteratorValidation(Consumer<GetShardIteratorRequest> validation) {
+            this.getShardIteratorValidation = validation;
+        }
+
+        @Override
+        public GetShardIteratorResponse getShardIterator(
+                GetShardIteratorRequest getShardIteratorRequest)
+                throws AwsServiceException, SdkClientException {
+            getShardIteratorValidation.accept(getShardIteratorRequest);
+            return GetShardIteratorResponse.builder().shardIterator(shardIterators.poll()).build();
+        }
+
+        public void setDescribeStreamResponse(List<DescribeStreamItem> items) {
+            describeStreamItems.addAll(items);
+        }
```

Comment:
Do we want `describeStreamValidation`?

Reply:
i don't think we require it here. We are only testing validation in request